cliquet/resource.py

```diff
@@ -123,7 +124,167 @@ def raise_invalid(self, location='body', **kwargs):
+            raise HTTPMethodNotAllowed()
+
+        self._raise_412_if_modified()
+
+        deleted = self.delete_records()
+
+        body = {
+            'items': deleted,
+        }
+
+        return body
+
+    @resource.view(permission='readonly', cors_headers=('Last-Modified',))
+    def get(self):
+        self._add_timestamp_header(self.request.response)
+        record = self.get_record(self.record_id)
+        self._raise_304_if_not_modified(record)
+        self._raise_412_if_modified(record)
+
+        return record
+
+    @resource.view(permission='readwrite')
+    def put(self):
+        try:
+            existing = self.get_record(self.record_id)
+            self._raise_412_if_modified(existing)
+        except HTTPNotFound:
+            existing = None
+
+        new_record = self.request.validated
+
+        new_id = new_record.setdefault(self.id_field, self.record_id)
+        if new_id != self.record_id:
+            error_msg = 'Record id does not match existing record'
+            self.raise_invalid(name=self.id_field, description=error_msg)
+
+        new_record = self.process_record(new_record, old=existing)
+        record = self.update_record(new_record, existing)
+        return record
+
+    @resource.view(permission='readwrite')
+    def patch(self):
+        record = self.get_record(self.record_id)
+        self._raise_412_if_modified(record)
+
+        changes = self.request.json
+
+        updated = self.apply_changes(record, changes=changes)
+
+        updated = self.process_record(updated, old=record)
+
+        nothing_changed = not any([record.get(k) != updated.get(k)
+                                   for k in changes.keys()])
+        if nothing_changed:
+            return record
+
+        record = self.update_record(updated, old=record)
+        return record
+
+    @resource.view(permission='readwrite')
+    def delete(self):
+        record = self.get_record(self.record_id)
+        self._raise_412_if_modified(record)
+
+        deleted = self.delete_record(record)
+        return deleted
+
+    #
+    # Storage
+    #
+
+    def get_records(self):
+        """Fetch the collection records, using querystring arguments for
+        sorting, filtering and pagination.
+
+        Override to implement custom querystring parsing, or post-process
+        records after their retrieval from storage.
+
+        :raises: :class:`pyramid.httpexceptions.HTTPBadRequest` if filters or
+            sorting are invalid.
+        :returns: A tuple with the list of records in the current page,
+            the total number of records in the result set, and the next page
+            url.
+        :rtype: tuple
+        """
+        filters = self._extract_filters()
+        sorting = self._extract_sorting()
+        pagination_rules, limit = self._extract_pagination_rules_from_token(
+            sorting)
+
+        include_deleted = self.modified_field in [f[0] for f in filters]
```

Comment:
I'm not totally convinced of having filters extraction here. If we mean to keep protocol stuff out of customization, this should probably go back to the endpoint code.

```python
def get_records(self, filters, sorting, pagination, limit, include_deleted):
       # Customize : disable sorting on certain fields
       sorting.remove(('age', 1))
       return super(..., ...).get_records(...)
```